merkle/hashers/tree_hasher.go

```diff
@@ -42,3 +48,47 @@ type MapHasher interface {
 	// TODO(gbelvin): Replace Size() with BitLength().
 	Size() int
 }
+
+var (
+	maxHashers = trillian.HashStrategy(len(trillian.HashStrategy_value))
+	logHashers = make([]LogHasher, maxHashers)
+	mapHashers = make([]MapHasher, maxHashers)
+)
+
+// RegisterLogHasher registers a hasher for use.
+func RegisterLogHasher(h trillian.HashStrategy, f LogHasher) {
```

Comment:
This solution assumes a hasher is a function of HashStrategy, but in fact it's a mix of HashAlgorithm and HashStrategy. We had this right on the (now deleted) trees functions.

I don't mind about the interface changes, but I'd strongly prefer if you kept `trees.Hash` and `trees.XHasher` as they were. I don't think we need a registry for hashers either; we've only got a few (hand-coded) of them anyway.